DMTN-148.tex

```diff
@@ -50,9 +50,276 @@
+\verb|obs_CAMERA_data| package implement one-off mapper methods to allow the
+product to be found.  However, only \verb|DEFECTS|, \verb|QE_CURVE|, and \verb|LINEARITY|
+follow this system, leaving the remaining products to be handled with
+camera-specific code and cameraGeom entries.  In addition to having
+diverse implementations, these handlers are often rigid, and cannot be
+versioned easily.
+
+\section{Transitioning to Gen3/What We Should Expect From Gen3}
+
+\subsection{A CalibrationBaseClass}
+The first step to a consistent calibration system is to ensure that
+all products in both generations use the \verb|DEFECTS|-like calibration
+product system.  For FITS image products, this requires future
+calibrations to be constructed with all of the header keyword values
+listed above.
+
+The remaining unconverted calibration types (\verb|CROSSTALK|, \verb|BFKERNEL|)
+should be converted to use \verb|DEFECTS|-like class, most likely by defining
+a CalibrationBaseClass, and making all of the calibration products
+subclasses of that.  This would enforce that all calibrations can be
+relied upon to provide getMetadata, setMetadata, readFits, writeFits,
+readText, and writeText methods.  Gen2 Butler instances would then be
+able to access all calibration products in a single consistent manner.
+This also ensures that Gen2 and Gen3 can use the same calibration
+products.
+
+\subsection{Metadata and Header Keywords}
+
+The required header keywords are currently poorly standarized.  The
+\verb|OBSTYPE| keyword is universally populated in all calibrations,
+regardless of if they are generated by constructCalibs.py or are
+curated by the \verb|obs_CAMERA| packages.  Curated calibrations also
+require the \verb|DETECTOR| and \verb|CALIBDATE|, the last of which is
+also populated by constructCalibs.py.  The \verb|CALIB_ID| is set by
+constructCalibs.py, and exists for the majority of curated
+calibrations, but is not consistently used.
+
+The keywords listed in table \ref{tab:metadata} are the proposed list
+of header keywords that will be required in the new operational
+calibrations in Gen3.  The majority of these are inserted at creation,
+identifying the content and applicable data range for the calibration.
+Only upon certification (the replacement for the Gen2 ingest) are the
+valid date ranges set.  To conform with the expected changes to FITS
+persistence in RFC-686, these and all other calibration-specific
+metadata fields will be upper-case.
+
+\begin{tabular}{l l l}
+  \label{tab:metadata}
+  Keyword & Inserted & Description \\
+  \hline
+  \verb|OBSTYPE| & Creation & Calibration dataset type contained in this file. \\
+  \verb|{OBSTYPE}_SCHEMA| & Creation & Storage schema for calibration.  \\
+  \verb|{OBSTYPE}_VERSION| & Creation & Storage schema version. \\
+  \verb|INSTRUME| & Creation & Instrument/camera this calibration belongs to. \\
+  \verb|DETECTOR| & Creation & Integer identifier for the ccd/detector. \\
```

Comment:
Is this best as an integer or as a per-camera-interpreted string?

Reply:
Detector number is the primary way of addressing a detector. Of course in the _data packages the directory tree uses detector name.

Reply:
I'll be updating this section to include the as-implemented changes, but `DETECTOR` is the integer `detectorId` that is used to access the appropriate element in the `afw.cameraGeom.Camera` object.  `DET_NAME` contains the detector name ("RXX_SYY" for `obs_lsst`-like camera), `RAFTNAME` and `SLOTNAME` the appropriate values (or None), and `DET_SER` containing the detector serial number.  Most of these will be informational only, but I'm anticipating that we'll want to add calibration validation in the future (particularly when we start exporting products to external butlers), and these will give ways to be as strict as needed in that validation.  In this context, validation means ensuring that the calibrations we have are appropriate for the exposure we're planning on applying them to.